Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+///
+/// ```swift
+/// var server = Server()
+///
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessServerTransport()
+/// server.transports.add(inProcessTransport)
+///
+/// // Create and register the 'Greeter' and 'Echo' services.
+/// server.services.register(GreeterService())
+/// server.services.register(EchoService())
+///
+/// // Create and add some interceptors.
+/// server.interceptors.add(StatsRecordingServerInterceptors())
+/// ```
+///
+/// ## Starting and stopping the server
+///
+/// Once you have configured the server call ``run()`` to start it. Calling ``run()`` starts each
+/// of the servers transports. A ``ServerError`` is thrown if any of the transports can't be
+/// started.
+///
+/// ```swift
+/// // Start running the server.
+/// try await server.run()
+/// ```
+///
+/// The ``run()`` method won't return until the server has finished handling all requests. You can
+/// signal to the server that it should stop accepting new requests by calling ``stopListening()``.
+/// This allows the server to drain existing requests gracefully. To stop the server more abruptly
+/// you can cancel the task running your server. If your services require additional resources that
+/// need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct Server: Sendable {
+  typealias Stream = RPCStream<ServerTransport.Inbound, ServerTransport.Outbound>
+
+  /// A collection of ``ServerTransport`` implementations that the server uses to listen
+  /// for new requests.
+  public var transports: Transports
+
+  /// The services registered which the server is serving.
+  public var services: Services
+
+  /// A collection of ``ServerInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request received
+  /// from the client will first be intercepted by the first added interceptor followed by the
+  /// second, and so on.
+  public var interceptors: Interceptors
+
+  /// The state of the server.
+  private enum State {
+    /// The server hasn't been started yet. Can transition to `starting` or `stopped`.
+    case notStarted
+    /// The server is starting but isn't accepting requests yet. Can transition to `running`
+    /// and `stopping`.
+    case starting
+    /// The server is running and accepting RPCs. Can transition to `stopping`.
+    case running
+    /// The server is stopping and no new RPCs will be accepted. Existing RPCs may run to
+    /// completion. May transition to `stopped`.
+    case stopping
+    /// The server has stopped, no RPCs are in flight and no more will be accepted. This state
+    /// is terminal.
+    case stopped
+  }
+
+  private let state: LockedValueBox<State>
```

Comment:
Interesting. I didn't expect to have a struct with reference semantics in the end. For `ServiceLifecycle` we decided to use an actor for the `ServiceGroup`. Here I would have actually expected a `class`. Especially since you can do weird things like this
```
var server = Server()
server.transport = [SomeTransport()]
async let _ = server.run()
server.transports = [OtherTransport()]
async let _ = server.run()
```

IMO here having a class for the `Server` makes more sense.

Reply:
Yeah, probably should be a `class` tbh. I started with a `struct` and didn't think to back out that change when I added the state management.